runtime/runc_jailer.go

```diff
@@ -0,0 +1,370 @@
+func (j runcJailer) BuildLinkFifoHandler() firecracker.Handler {
+	return firecracker.Handler{
+		Name: jailerFifoHandlerName,
+		Fn: func(ctx context.Context, m *firecracker.Machine) error {
+			contentsPath := j.RootPath()
+			fifoFileName := filepath.Base(m.Cfg.LogFifo)
+			newFifoPath := filepath.Join(contentsPath, fifoFileName)
+			if err := os.Link(m.Cfg.LogFifo, newFifoPath); err != nil {
+				return err
+			}
+			m.Cfg.LogFifo = newFifoPath
+
+			metricFifoFileName := filepath.Base(m.Cfg.MetricsFifo)
+			newMetricFifoPath := filepath.Join(contentsPath, metricFifoFileName)
+			if err := os.Link(m.Cfg.MetricsFifo, newMetricFifoPath); err != nil {
+				return err
+			}
+			m.Cfg.MetricsFifo = newMetricFifoPath
+
+			return nil
+		},
+	}
+}
+
+// ExposeDeviceToJail will inspect the given file, srcFilePath, and based on the
+// file type, proper handling will occur to ensure that the file is visible in
+// the jail.
+func (j runcJailer) ExposeDeviceToJail(srcMountPath string) error {
+	uid := j.uid
+	gid := j.gid
+
+	stat := syscall.Stat_t{}
+	if err := syscall.Stat(srcMountPath, &stat); err != nil {
+		return err
+	}
+
+	path := j.RootPath()
+	dst := ""
+	if (stat.Mode & syscall.S_IFMT) == syscall.S_IFBLK {
+		path = filepath.Join(path, "dev")
+		dst = filepath.Join(path, filepath.Base(srcMountPath))
+		if err := exposeBlockDeviceToJail(dst, int(stat.Rdev), int(uid), int(gid)); err != nil {
+			return err
+		}
+	} else {
+		return fmt.Errorf("unsupported mode: %v", stat.Mode)
+	}
+
+	return nil
+}
+
+// exposeBlockDeviceToJail will call mknod on the block device to ensure
+// visibility of the device
+func exposeBlockDeviceToJail(dst string, rdev, uid, gid int) error {
+	if err := syscall.Mknod(dst, syscall.S_IFBLK, rdev); err != nil {
+		return err
+	}
+
+	if err := os.Chmod(dst, 0600); err != nil {
+		return err
+	}
+
+	if err := os.Chown(dst, uid, gid); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func copyFile(src, dst string, mode os.FileMode) error {
+	srcFile, err := os.Open(src)
+	if err != nil {
+		return errors.Wrapf(err, "failed to open %v", src)
+	}
+	defer srcFile.Close()
+
+	info, err := os.Stat(dst)
+	switch {
+	case err == nil:
+		// no error indicates the path already exists, we check to see if it is a
+		// directory we will concatenate the file name provided in source.
+		if info.IsDir() {
+			dst = filepath.Join(dst, filepath.Base(src))
+		}
+	case os.IsNotExist(err):
+		// this occurs when a full path is provided including the file name
+	default:
+		return err
+	}
+
+	dstFile, err := os.OpenFile(dst, os.O_CREATE|os.O_WRONLY|os.O_EXCL, mode)
+	if err != nil {
+		return errors.Wrapf(err, "failed to open %v", dstFile)
+	}
+	defer dstFile.Close()
+
+	_, err = io.Copy(dstFile, srcFile)
+	if err != nil {
+		return errors.Wrap(err, "failed to copy to destination")
+	}
+	return nil
+}
+
+var (
+	stdout *os.File
+	stderr *os.File
+)
+
+func init() {
```

Comment:
This `init` is *way buried*!

`init` functions can be very surprising as they'll run as soon as the package is imported and are not super obvious to find (they avoid the normal control flow through a program).  I think we probably *don't* want this function to exist at all (it's creating some constant filenames with "test" in the name).  If we do ultimately want some init function like this, it needs to be at the absolute top of the file and declared in the package-level godoc as something that happens at package-import time.